content/blog/22-23/reverse_engineering_epaper.md

```diff
@@ -0,0 +1,141 @@
+
+<%= figure 'https://pics.zeus.gent/lxmAPEmHlbodZm4IV4RWYzRSaAKtlXomsGySQdhz.png', 'Pinout of the CC2510' %>
+
+After inspecting the pinout, one pin, namely the DCOUPL pin, stood out:
+
+> DCOUPL: Power decoupling: 1.8 V digital power supply decoupling
+
+The chip itself runs on 3.3V and no external pins have 1.8V logic level, so at first glance, it would be a bit strange for the chip to have a 1.8V power supply. However, this is actually rather common: the internal logic in the microcontroller (the CPU, RAM, flash, ...) very likely uses 1.8V. The DCOUPL pin is meant to be attached to an external capacitor to smooth out the internal 1.8V power supply. For us, this is handy, because now we have a direct connection to the internal power supply.
+
+To glich the microcontroller, we do some micro-surgery: we remove the decoupling capacitor and attach a fast MOSFET (a digital switch) to the DCOUPL and ground pin. When the MOSFET is enabled, the 1.8V power supply is shorted to ground, and the chip glitches. Succesfully mounting a voltage glitching attack then becomes a matter of correctly timing the glitch and closing the MOSFET for the right amount of time. Those timings need to be very precise, in the order of nanosecond precision. We basically need to pull a pin high for a very short time, at a very precise time in the debug sequence.
+
+<%= figure 'https://pics.zeus.gent/Qgbby9p0w4cIiOTUwh6Yg0d1hhihzDLXpgj4PnLN.jpg', 'Closeup of the microcontroller. The DCOUPL capacitor we need to remove is marked in orange.' %>
+
+<%= figure 'https://pics.zeus.gent/uEuYSYzOttjkZhwtLoH8p3mOPbkkiLPwcLbdlraN.jpg', 'Closeup of the microcontroller after the MOSFET was added to the DCOUPL pin' %>
+
+
+We previously used the Raspberry Pi Pico to communicate with the board; for the glitch timings, we'll use the PIO (programmable IO) feature of the RP2040 chip on the Pi Pico. This cool piece of hardware allows us to set pins at the clock speed of the chip: so we can have a 125 MHz signal. We can feed the PIO peripheral using DMA (Direct Memory Access), so we can do other things while the glitch signal control runs in the background. The buffer that is used for the glich pin is thus filled with zeroes, except at one location, where there are a couple of consecutive ones.
+
+A couple of modifications onto this idea were needed to make the attack more reliable:
+
+- Overclocking the Pi Pico to 250 MHz instead of the default 125 MHz. This doubles the precision in glitch length and duration.
+- The Pi Pico does it's serial over software USB. The interrupts from USB sometimes throw of the timings, so we run the USB stack on core 0, and our code on core 1, with interrupts disabled.
+- We set the drive strenght of the power pin to 12 mA, and the slew rate to 'fast' (instead of the default 'slow'). This makes sure that the internal capacitance of the MOSFET gate charges sufficiently fast, so we have a nice sharp edge on the glitch pin.
```

Comment:
```suggestion
- We set the drive strength of the power pin to 12 mA, and the slew rate to 'fast' (instead of the default 'slow'). This makes sure that the internal capacitance of the MOSFET gate charges sufficiently fast, so we have a nice sharp edge on the glitch pin.
```